src/vcf/fasta.cpp

```diff
@@ -0,0 +1,131 @@
+/**
+ * Copyright 2018 EMBL - European Bioinformatics Institute
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <memory>
+
+#include <boost/algorithm/string/predicate.hpp>
+#include <boost/iostreams/copy.hpp>
+
+#include "bioio/bioio.hpp"
+
+#include "util/file_utils.hpp"
+#include "util/logger.hpp"
+#include "util/stream_utils.hpp"
+#include "util/string_utils.hpp"
+#include "vcf/compression.hpp"
+#include "vcf/fasta.hpp"
+#include "vcf/string_constants.hpp"
+
+std::shared_ptr<std::istream> create_gunzipped_stream(std::shared_ptr<std::istream> stream) {
+  std::shared_ptr<std::stringstream> ss(new std::stringstream());
+  try {
+    boost::iostreams::filtering_istream decompressedStream;
+    ebi::vcf::create_uncompressed_stream(*stream, ebi::vcf::GZ, decompressedStream);
+    boost::iostreams::copy(decompressedStream, *ss);
+  } catch (std::exception const &ex) {
+    BOOST_LOG_TRIVIAL(error) << ex.what();
+  }
+
+  return std::static_pointer_cast<std::istream>(ss);
+}
+
+std::shared_ptr<std::istream> create_decompressed_stream(std::shared_ptr<std::istream> stream) {
+  std::vector<char> line;
+  ebi::vcf::get_magic_num(*stream, line);
+  std::string file_extension = ebi::vcf::get_compression("", line);
+
+  if ( file_extension == ebi::vcf::GZ ) {
```

Comment:
I am afraid if uncompressing a fasta file would allow random reads that we do in `bioio` library

Reply:
If you are referring to the current implementation of `bioio` library I did not see it could handle compressed fasta file/stream. If have missed something, please point it out @srbcheema1. The point of decompressing `gz` stream first is to allow random access later, which by its nature a `gz` stream does not allow.

Reply:
ok, maybe I'm not aware of all the conversation but now I realised this PR adds 3 features (compressed fastas, remote fastas, ENA API), while the jira ticket asks only for the ENA API. Did I miss anything?

Reply:
because I can see that the design was changed to load even decompressed fastas in memory. The point about having fasta indexes is to not load the whole fasta in memory. have you tried validating a vcf using a big decompressed fasta?

Reply:
The new implementation has removed the functionalities of reading compressed or remote fasta files. It is now only getting reference sequences from ENA API.